packages/web-frontend/cypress/scripts/generateConfig/VisualisationFilter.js

```diff
@@ -0,0 +1,93 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { keyBy, upperFirst } from 'lodash';
+
+import { getUniqueEntries, hasIntersection, toArray } from '@tupaia/utils';
+
+class VisualisationFilter {
+  constructor(db, filter) {
+    this.db = db;
+    this.filter = filter;
+  }
+
+  /**
+   * @abstract
+   */
+  // eslint-disable-next-line no-unused-vars
+  fetchRecords = async objectUrls => {
+    throw new Error('Any subclass of VisualisationFilter must implement the "fetchRecords" method');
+  };
+
+  /**
+   * @abstract
+   */
+  getBuilderKey = () => {
+    throw new Error(
+      'Any subclass of VisualisationFilter must implement the "getBuilderKey" method',
+    );
+  };
+
+  /**
+   * @protected
+   */
+  getName = () => 'visualisation';
+
+  apply = async objectUrls => {
+    if (Object.keys(this.filter).length === 0) {
+      return objectUrls;
+    }
+
+    const records = await this.fetchRecords(objectUrls);
+    const recordsById = keyBy(records, 'id');
+
+    const filterUrl = url =>
+      Object.entries(this.filter).some(([key, targetValue]) => {
+        const record = recordsById[url.id];
+        if (!record) {
+          throw new Error(
+            `${upperFirst(this.getName())} with id "${url.id}" was not found in the database`,
+          );
+        }
+
+        const isBuilderFilter = key === this.getBuilderKey();
+        const value = isBuilderFilter ? this.getBuildersInRecord(recordsById[url.id]) : url[key];
+        return hasIntersection(toArray(targetValue), toArray(value));
+      });
+
+    return objectUrls.filter(filterUrl);
+  };
+
+  getBuildersInRecord = record => {
+    const builderKey = this.getBuilderKey();
+    const nestedBuilders = Object.values(record[`${builderKey}Config`][`${builderKey}s`] || {}).map(
+      nestedConfig => nestedConfig[builderKey],
+    );
+    return getUniqueEntries([record[builderKey], ...nestedBuilders]);
+  };
+}
+
+export class DashboardReportFilter extends VisualisationFilter {
+  fetchRecords = async objectUrls =>
+    this.db.find('dashboardReport', {
+      id: objectUrls.map(u => u.id),
+      drillDownLevel: null, // Drill down reports are not currently supported in e2e tests
+    });
+
+  getBuilderKey = () => 'dataBuilder';
+
+  getName = () => 'dashboard report';
+}
+
+export class MapOverlayFilter extends VisualisationFilter {
+  fetchRecords = async objectUrls =>
+    this.db.find('mapOverlay', {
+      id: objectUrls.map(u => u.id).flat(),
+    });
+
+  getBuilderKey = () => 'measureBuilder';
+
+  getName = () => 'map overlay';
+}
```

Comment:
Yeah things got a bit simpler with https://github.com/beyondessential/tupaia/pull/2781 - we decided that the abstract getters may be an overkill in the JS world, so hopefully now the shared code is easier to read!

I am open to a different approach though - in fact, lately I've been trying to focus on simplicity rather than abstraction. I try to see this from a practical standpoint: what makes code really easier to maintain? Sometimes, having to update code in two different places is quicker and easier than understanding a single piece of abstraction that is hard to understand. The only drawback with the simpler approach is that it may be harder for people to track all the places where an update should take place, but that's what PR reviews/manual/automated testing is for.

Reply:
I've come around big time in the same way! You put that very well btw :-)